subs/pantry/src/Pantry/Archive.hs

```diff
@@ -387,41 +393,60 @@ parseArchive pli archive fp = do
             case Map.lookup (seSource se) blobs of
               Nothing -> error $ "Impossible: blob not found for: " ++ seSource se
               Just blobKey -> pure (sfp, TreeEntry blobKey (seType se))
-
           -- parse the cabal file and ensure it has the right name
-          (cabalPath, cabalEntry@(TreeEntry cabalBlobKey _)) <- findCabalFile pli tree
+          (cabalPath, cabalEntry@(TreeEntry cabalBlobKey _), btype) <- findCabalOrHpackFile pli tree
           mbs <- withStorage $ loadBlob cabalBlobKey
           bs <-
             case mbs of
               Nothing -> throwIO $ TreeReferencesMissingBlob pli cabalPath cabalBlobKey
               Just bs -> pure bs
-          (_warnings, gpd) <- rawParseGPD (Left pli) bs
+          cabalBs <- if btype == CabalFile
+                     then return bs
+                     else do
+                       getCabalBS bs
+          (_warnings, gpd) <- rawParseGPD (Left pli) cabalBs
           let ident@(PackageIdentifier name _) = package $ packageDescription gpd
-          when (cabalPath /= cabalFileName name) $
+          when (cabalPath /= cabalFileName name && btype == CabalFile) $
             throwIO $ WrongCabalFileName pli cabalPath name
 
           -- It's good! Store the tree, let's bounce
-          (_tid, treeKey) <- withStorage $ storeTree ident tree cabalEntry
+          (_tid, treeKey) <- withStorage $ storeTree ident tree cabalEntry btype
+          let tentry = if btype == CabalFile
+                       then PCCabalFile cabalEntry
+                       else PCHpack cabalEntry
           pure Package
             { packageTreeKey = treeKey
             , packageTree = tree
-            , packageCabalEntry = cabalEntry
+            , packageCabalEntry = tentry
             , packageIdent = ident
             }
 
-findCabalFile
+findCabalOrHpackFile
   :: MonadThrow m
   => PackageLocationImmutable -- ^ for exceptions
   -> Tree
-  -> m (SafeFilePath, TreeEntry)
-findCabalFile loc (TreeMap m) = do
-  let isCabalFile (sfp, _) =
+  -> m (SafeFilePath, TreeEntry, BuildFileType)
+findCabalOrHpackFile loc (TreeMap m) = do
+  let isCabalOrHpackFile (sfp, _) =
+        let txt = unSafeFilePath sfp
+         in not ("/" `T.isInfixOf` txt) && (".cabal" `T.isSuffixOf` txt || "package.yaml" == txt)
+      isCabalFile (sfp, _) =
         let txt = unSafeFilePath sfp
-         in not ("/" `T.isInfixOf` txt) && ".cabal" `T.isSuffixOf` txt
-  case filter isCabalFile $ Map.toList m of
+         in not ("/" `T.isInfixOf` txt) && (".cabal" `T.isSuffixOf` txt)
+      isHpackFile (sfp, _) =
+        let txt = unSafeFilePath sfp
+         in not ("/" `T.isInfixOf` txt) && ("package.yaml" == txt)
```

Comment:
how could we have "/" in `txt` being equal to `"package.yaml"`?

Reply:
Fixed.